tests/SchemaTest.php

```
@@ -0,0 +1,100 @@
+    }
+
+    /**
+     * @test
+     * @dataProvider validFileProvider
+     */
+    public function valid_documents_pass(DOMDocument $document, string $schema) : void
+    {
+        $result = $document->relaxNGValidate($schema);
+
+        $this->assertTrue($result);
+    }
+
+    /**
+     * @test
+     * @dataProvider invalidFileProvider
+     */
+    public function invalid_documents_fail(DOMDocument $document, string $schema, array $expected) : void
+    {
+        $document->relaxNGValidate($schema);
+
+        $actual = map(
+            libxml_get_errors(),
+            function (LibXMLError $error) : array {
+                return [
+                    'line' => $error->line,
+                    'message' => trim($error->message),
+                ];
+            }
+        );
+
+        $this->assertSame($expected, $actual);
+    }
+
+    public function validFileProvider() : iterable
+    {
+        $files = Finder::create()->files()
+            ->name('*.xml')
+            ->in(__DIR__)
+            ->path('~/valid/~');
+
+        return $this->extractSchemas($files);
+    }
+
+    public function invalidFileProvider() : iterable
+    {
+        $files = Finder::create()->files()
+            ->name('*.xml')
+            ->in(__DIR__)
+            ->path('~/invalid/~');
+
+        return $this->extractSchemas($files);
+    }
+
+    private function extractSchemas(Finder $files) : iterable
+    {
+        foreach ($files as $file) {
+            $dom = FluentDOM::load($file->getContents());
+
+            $schema = $dom('substring-before(substring-after(/processing-instruction("xml-model"), \'"\'), \'"\')');
+            $schema = "{$file->getPath()}/{$schema}";
+
+            $expectedFailures = map(
+                $dom('/processing-instruction("expected-error")'),
+                function (ProcessingInstruction $instruction) : array {
+                    preg_match('~line="([0-9]+)"\s+message="([^"]*?)"~', $instruction->nodeValue, $matches);
```

needs a check that `preg_match` succeeds: any string may come from `$instruction->nodeValue`

---

ah, parsing XML with a regular expression?

---

No, processing instructions are just strings, but it's common to make them look attribute like. This will fail hard if the string is invalid, which will just fails the tests.

---

The only value exposed here is the internal part of the Processing Instruction, so it should be regular given the examples containing no `"`

---

(so only needs a check for success)

---

Primarily because an error message like `you mistyped a processing instruction` is friendlier than `undefined index 1`